File: python/src/compute_cfo/exporters.py
"""Cost event exporters."""

from __future__ import annotations

import json
import sys
import urllib.request
from typing import Callable

from .types import CostEvent

Exporter = Callable[[CostEvent], None]


def console_exporter(event: CostEvent) -> None:
    """Print cost event to stderr."""
    cost_str = f"${event.cost_usd:.6f}" if event.cost_usd is not None else "$?.??????"
    total_tokens = event.input_tokens + event.output_tokens

    parts = [
        f"\033[90m[compute-cfo]\033[0m",
        f"\033[1m{event.model}\033[0m",
        f"{total_tokens} tokens",
        f"\033[33m{cost_str}\033[0m",
    ]

    if event.tags:
        tag_str = " ".join(f"{k}:{v}" for k, v in event.tags.items())
        parts.append(f"\033[90m{tag_str}\033[0m")

    print(" | ".join(parts), file=sys.stderr)


def jsonl_exporter(path: str = "compute_cfo_events.jsonl") -> Exporter:
    """Return an exporter that appends JSON lines to a file."""

    def _export(event: CostEvent) -> None:
        with open(path, "a") as f:
            f.write(json.dumps(event.to_dict()) + "\n")

    return _export


def webhook_exporter(url: str) -> Exporter:
    """Return an exporter that POSTs cost events to a URL."""

    def _export(event: CostEvent) -> None:
        data = json.dumps(event.to_dict()).encode("utf-8")
        req = urllib.request.Request(
            url,
            data=data,
            headers={"Content-Type": "application/json"},
            method="POST",
        )
        try:
            urllib.request.urlopen(req, timeout=5)
        except Exception:
            pass  # fire-and-forget

    return _export


def get_exporter(spec: str) -> Exporter:
    """Parse an exporter specification string and return the exporter.

    Supported formats:
        "console" — print to stderr
        "jsonl" — write to compute_cfo_events.jsonl
        "jsonl:/path/to/file.jsonl" — write to custom path
        "webhook:https://example.com/hook" — POST to URL
    """
    if spec == "console":
        return console_exporter

    if spec == "jsonl":
        return jsonl_exporter()

    if spec.startswith("jsonl:"):
        return jsonl_exporter(spec[6:])

    if spec.startswith("webhook:"):
        return webhook_exporter(spec[8:])

    raise ValueError(
        f"Unknown exporter: {spec!r}. "
        f"Use 'console', 'jsonl', 'jsonl:/path', or 'webhook:URL'."
    )

File: typescript/src/exporters.ts
/**
 * Cost event exporters.
 */

import * as fs from 'fs';
import { CostEvent, costEventToDict } from './types';

export type Exporter = (event: CostEvent) => void;

export function consoleExporter(event: CostEvent): void {
  const costStr =
    event.costUsd !== null ? `$${event.costUsd.toFixed(6)}` : '$?.??????';
  const totalTokens = event.inputTokens + event.outputTokens;

  const parts: string[] = [
    `[compute-cfo]`,
    event.model,
    `${totalTokens} tokens`,
    costStr,
  ];

  if (Object.keys(event.tags).length > 0) {
    const tagStr = Object.entries(event.tags)
      .map(([k, v]) => `${k}:${v}`)
      .join(' ');
    parts.push(tagStr);
  }

  console.error(parts.join(' | '));
}

export function jsonlExporter(
  path: string = 'compute_cfo_events.jsonl'
): Exporter {
  return (event: CostEvent) => {
    fs.appendFileSync(path, JSON.stringify(costEventToDict(event)) + '\n');
  };
}

export function webhookExporter(url: string): Exporter {
  return (event: CostEvent) => {
    const data = JSON.stringify(costEventToDict(event));
    fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: data,
    }).catch(() => {}); // fire-and-forget
  };
}

export function getExporter(spec: string): Exporter {
  if (spec === 'console') return consoleExporter;
  if (spec === 'jsonl') return jsonlExporter();
  if (spec.startsWith('jsonl:')) return jsonlExporter(spec.slice(6));
  if (spec.startsWith('webhook:')) return webhookExporter(spec.slice(8));
  throw new Error(
    `Unknown exporter: "${spec}". Use 'console', 'jsonl', 'jsonl:/path', or 'webhook:URL'.`
  );
}
